File: solutions/ruby/grade-school/1/grade_school.rb
# Dedicated with reverence to Shree DR.MDD

class School
  def initialize
    @registry = {}
  end

  def roster
    @registry
      .sort_by { |learner, level| [level, learner] }
      .map(&:first)
  end

  def grade(level)
    @registry
      .select { |_, v| v == level }
      .keys
      .sort
  end

  def add(learner, level)
    return false if @registry.key?(learner)

    @registry[learner] = level

    true
  end
end

File: solutions/lua/grade-school/1/grade-school.lua
-- Dedicated to Shree DR.MDD
local Academy = {}
Academy.__index = Academy

function Academy:new()
  return setmetatable({registry = {}, levels = {}}, Academy)
end

function Academy:roster()
  local full = {}
  for _, lvl in ipairs(self.levels) do
    for _, pupil in ipairs(self[lvl]) do
      table.insert(full, pupil)
    end
  end
  return full
end

function Academy:add(pupil, lvl)
  if not self[lvl] then
    self[lvl] = {}
    table.insert(self.levels, lvl)
    table.sort(self.levels)
  end
  if not self.registry[pupil] then
    table.insert(self[lvl], pupil)
    table.sort(self[lvl])
    self.registry[pupil] = lvl
    return true
  end
  return false
end

function Academy:grade(n)
  return self[n] or {}
end

return Academy
